### PR TITLE
Embedded omnisci integration and arrow dataframes

### PR DESCRIPTION
**Pull request by anton-malakhov**
Replacing https://github.com/intel-go/modin/pull/57 for integration into omnisci_on_ray branch

## What do these changes do?
integrates DBE and Arrow tables into Modin along with read_csv functionality

- [+] passes `flake8 modin` (not everything though)
- [ ] passes `black --check modin`
- [ ] signed commit with `git commit -s` 
- [ ] Resolves #? 
- [ ] tests added and passing

### CODE REVIEW

File: modin/engines/base/frame/data.py [hunk 1]
```diff
@@ -1339,6 +1364,7 @@ def to_numpy(self):
         Returns:
             NumPy array.
         """
+        raise "to_numpy"
```
Comment (ienkovich) [inline modin/engines/base/frame/data.py:1367]: Debug code?

File: modin/engines/base/frame/partition_manager.py [hunk 2]
```diff
@@ -316,6 +316,26 @@ def from_pandas(cls, df, return_dims=False):
             ]
             return np.array(parts), row_lengths, col_widths
 
+    @classmethod
+    def from_arrow(cls, at, return_dims=False):
+        put_func = cls._partition_class.put_arrow
+
+        parts = [[put_func(at)]]
+        #TODO? [
```
Comment (ienkovich) [inline modin/engines/base/frame/partition_manager.py:324]: Let's remove all these comments. If there are real TODOs here, then make an issue and reference it.

File: modin/experimental/engines/omnisci_on_ray/frame/data.py [hunk 3]
```diff
@@ -91,7 +91,7 @@ def __init__(
         else:
             self._dtypes = dtypes
 
-        if partitions is not None:
+        if partitions is not None and self._index_cols is not None:
```
Comment (ienkovich) [inline modin/experimental/engines/omnisci_on_ray/frame/data.py:94]: Why is this changed?

File: modin/experimental/engines/omnisci_on_ray/frame/partition_manager.py [hunk 4]
```diff
@@ -68,40 +72,43 @@ def run_exec_plan(cls, plan, index_cols, dtypes):
...
         calcite_json = CalciteSerializer().serialize(calcite_plan)
 
-        sql = "execute relalg " + calcite_json
-        df = OmnisciServer()._worker._conn._execute(sql).to_df()
+        curs = omniSession.executeRA("execute relalg " + calcite_json)
+        assert curs
+        rb = curs.getArrowRecordBatch()
+        assert rb
+        at = pyarrow.Table.from_batches([rb])
+        # df = rb.to_pandas()  # TODO continue with arrow frames #1721
```
Comment (ienkovich) [inline modin/experimental/engines/omnisci_on_ray/frame/partition_manager.py:90]: I think we can now clean-up this and following comments.

File: modin/experimental/engines/omnisci_on_ray/io.py [hunk 5]
```diff
@@ -14,9 +14,285 @@
...
+
+            from pyarrow.csv import read_csv, timestamp, ParseOptions, ConvertOptions, ReadOptions
+
+            column_types= dtype if dtype is dict else {}
+            if( (type(parse_dates) is list) # and (type(parse_dates[0]) is str)  # like parse_dates=["dd",]
+                and type(column_types) is dict):
+                for c in parse_dates:
+                    column_types[c] = timestamp('s') 
+
+            """
```
Comment (ienkovich) [inline modin/experimental/engines/omnisci_on_ray/io.py:149]: Why do you copy Arrow classes documentation here?

File: modin/experimental/engines/omnisci_on_ray/test/test_dataframe.py [hunk 6]
```diff
@@ -1,21 +1,18 @@
...
+import numpy as np
+import pytest
 
 os.environ["MODIN_EXPERIMENTAL"] = "True"
 os.environ["MODIN_ENGINE"] = "ray"
 os.environ["MODIN_BACKEND"] = "omnisci"
-
-if os.environ["OMNISCI_SERVER"] is None:
-    raise RuntimeError("OMNISCI_SERVER variable must be initialized")
+sys.setdlopenflags(1 | 256)  # RTLD_LAZY+RTLD_GLOBAL
```
Comment (ienkovich) [inline modin/experimental/engines/omnisci_on_ray/test/test_dataframe.py:10]: We already have required code in omnisci_worker.py

File: modin/pandas/io.py [hunk 7]
```diff
@@ -106,7 +106,8 @@ def parser_func(
         memory_map=False,
         float_precision=None,
     ):
-        _, _, _, kwargs = inspect.getargvalues(inspect.currentframe())
```
Comment (ienkovich) [inline modin/pandas/io.py:109]: Is it just refactoring? If yes, then it should go in a separate PR directly to master.

File: modin/experimental/engines/omnisci_on_ray/test/test_dataframe.py (hunk 6 above)
Comment (ienkovich) [reply modin/experimental/engines/omnisci_on_ray/test/test_dataframe.py:10]: Looks like this is the reason of 'free(): invalid pointer' issue

File: modin/experimental/engines/omnisci_on_ray/frame/data.py (hunk 3 above)
Comment (anton-malakhov) [reply modin/experimental/engines/omnisci_on_ray/frame/data.py:94]: possibly left here while investigating an issue

Comment (anton-malakhov) [reply modin/experimental/engines/omnisci_on_ray/frame/data.py:94]: hm.. used to prevent some segfault, will remove for now

File: modin/experimental/engines/omnisci_on_ray/test/test_dataframe.py (hunk 6 above)
Comment (anton-malakhov) [reply modin/experimental/engines/omnisci_on_ray/test/test_dataframe.py:10]: removed - works so far

File: modin/pandas/io.py (hunk 7 above)
Comment (anton-malakhov) [reply modin/pandas/io.py:109]: things depend on the following code - it's not just refactoring

Comment (Garra1980) [reply modin/pandas/io.py:109]: I don't think we can that easily change general parts

File: modin/experimental/engines/omnisci_on_ray/test/test_dataframe.py [hunk 8]
```diff
@@ -900,3 +936,7 @@ def test_cat_codes(self):
         pandas_df["a"] = pandas_df["a"].cat.codes
 
         df_equals(pandas_df, exp)
+
+
+if __name__ == "__main__":
```
Comment (Garra1980) [inline modin/experimental/engines/omnisci_on_ray/test/test_dataframe.py:941]: Why do we need this?

Comment (anton-malakhov) [reply modin/experimental/engines/omnisci_on_ray/test/test_dataframe.py:941]: easy path for running the test separately from other tests and under debugger

File: modin/pandas/io.py (hunk 7 above)
Comment (anton-malakhov) [reply modin/pandas/io.py:109]: hm.. where is my answer? I said that it is more than refactoring

File: modin/experimental/engines/omnisci_on_ray/test/test_dataframe.py [hunk 9]
```diff
@@ -36,6 +32,46 @@ def run_and_compare(fn, data, data2=None, *args, **kwargs):
...
+        """ check with the following arguments: names, dtype, skiprows, delimiter """
+        csv_file = os.path.join(self.root, "modin/pandas/test/data", "test_usecols.csv")
+
+        for kwargs in (
+            {"delimiter": ','}, {"sep": None},
+            {"skiprows": 1, "names": ["A", "B", "C", "D", "E"]},
+            {"dtype": {'a': 'int32', 'e': 'string'}},
+        ):
+            rp = pd.read_csv(csv_file, **kwargs)
+            rm = to_pandas(mpd.read_csv(csv_file, engine='arrow', **kwargs))
```
Comment (ienkovich) [inline modin/experimental/engines/omnisci_on_ray/test/test_dataframe.py:48]: Do we need to pass engine arg now to use arrow parser?

Comment (anton-malakhov) [reply modin/experimental/engines/omnisci_on_ray/test/test_dataframe.py:48]: no, but it forces to stop before switching to fail-safe rout.

File: modin/experimental/engines/omnisci_on_ray/frame/partition_manager.py [hunk 10]
```diff
@@ -68,40 +72,37 @@ def run_exec_plan(cls, plan, index_cols, dtypes):
...
-                df[col] = df[col].astype("category")
-
-        if index_cols is not None:
-            df = df.set_index(index_cols)
-            df.index.rename(cls._names_from_index_cols(index_cols), inplace=True)
+        # for col in dtypes.index:
+        #    if dtypes[col] == "bool":
+        #        df[col] = df[col].astype("bool")
+        #    elif dtypes[col].name == "category":
+        #        df[col] = df[col].astype("category")
```
Comment (anton-malakhov) [inline modin/experimental/engines/omnisci_on_ray/frame/partition_manager.py:98]: @ienkovich do we need this commented code here?

File: modin/experimental/engines/omnisci_on_ray/frame/calcite_serializer.py [hunk 11]
```diff
@@ -126,11 +125,11 @@ def serialize_literal(self, literal):
...
                 "literal": int(str_val.replace(".", "")),
                 "type": "DECIMAL",
-                "target_type": "DECIMAL",
+                "target_type": "DOUBLE",
                 "scale": scale,
                 "precision": precision,
-                "type_scale": scale,
-                "type_precision": precision,
+                "type_scale": -2147483648,
+                "type_precision": 15,
```
Comment (anton-malakhov) [inline modin/experimental/engines/omnisci_on_ray/frame/calcite_serializer.py:132]: @ienkovich is it yours?

File: modin/experimental/engines/omnisci_on_ray/io.py [hunk 12]
```diff
@@ -14,9 +14,172 @@
...
+from modin.experimental.engines.omnisci_on_ray.frame.omnisci_worker import OmnisciServer
 
 class OmnisciOnRayIO(RayIO):
 
     frame_cls = OmnisciOnRayFrame
     query_compiler_cls = DFAlgQueryCompiler
+
+    #@classmethod
+    #def from_arrow(cls, at):
+    #    return cls.query_compiler_cls.from_arrow(at, cls.frame_cls)
```
Comment (anton-malakhov) [inline modin/experimental/engines/omnisci_on_ray/io.py:27]: to be removed

File: modin/pandas/io.py (hunk 7 above)
Comment (anton-malakhov) [reply modin/pandas/io.py:109]: rolled back

File: modin/experimental/engines/omnisci_on_ray/frame/partition_manager.py (hunk 10 above)
Comment (ienkovich) [reply modin/experimental/engines/omnisci_on_ray/frame/partition_manager.py:98]: no
